tiled/adapters/resource_cache.py

```diff
@@ -0,0 +1,61 @@
+import os
+from typing import Callable, Optional
+
+import cachetools
+
+# Cached items will be evicted if not used for a specified time interval
+# ("time to use"). If the cache size reaches its max size, the least recently
+# used cache item will be evicted.
+DEFAULT_MAX_SIZE = int(os.getenv("TILED_RESOURCE_CACHE_MAX_SIZE", "1024"))
+DEFAULT_TIME_TO_USE_SECONDS = float(os.getenv("TILED_RESOURCE_CACHE_TTU", "60."))
+
+_cache = None
+
+
+def get_resource_cache():
+    global _cache
+    if _cache is None:
+        cache = default_resource_cache()
+        set_resource_cache(cache)
+    return _cache
+
+
+def set_resource_cache(cache):
+    global _cache
+    _cache = cache
+
+
+def default_ttu(_key, value, now):
+    """
+    Retain cached items for at most 60 seconds.
+    """
+    return DEFAULT_TIME_TO_USE_SECONDS + now
+
+
+def default_resource_cache():
+    return cachetools.TLRUCache(DEFAULT_MAX_SIZE, default_ttu)
+
+
+def with_resource_cache(
+    cache_key: str,
+    factory: Callable,
+    *args,
+    _resource_cache: Optional[cachetools.Cache] = None,
+    **kwargs,
+):
+    """
+    Use value from cache or, if not present, call factory(*args, **kwargs) and cache result.
+    """
+    if _resource_cache is None:
+        cache = get_resource_cache()
+    else:
+        cache = _resource_cache
+    # Return cached value if found.
+    value = cache.get(cache_key)
```

Comment:
I couldn't easily find what happens to the `ttu` when an item is returned from the cache: is `ttu` updated for another 60 seconds or just let to expire? It doesn't matter that much, of course, especially for this PR, but could be interesting to see if some resources are requested so often that _recaching_ them creates a considerable overhead.

Reply:
Good thought.

The expiry time is computed once, at cache insertion, and not renewed.

https://github.com/tkem/cachetools/blob/d8b5c608273051f2ebddaec8a4d033c66369dd7d/src/cachetools/__init__.py#L546

For our use case it does seem better to let the expiry renew (the clock reset) on access. I cannot see anyway to make it renew using public APIs. Worth following up in a later PR, and maybe even a PR into cachetools.